Null pointer exception using Global Aggregation and then Nested Aggregation

<!-- Bug report -->

**Elasticsearch version** (`bin/elasticsearch --version`): 5.1.2

**Plugins installed**: []

**JVM version** (`java -version`):
openjdk version "1.8.0_151"
OpenJDK Runtime Environment (build 1.8.0_151-b12)
OpenJDK 64-Bit Server VM (build 25.151-b12, mixed mode)

**OS version** (`uname -a` if on a Unix-like system):

Linux xxxx 2.6.32-642.el6.x86_64 #1 SMP Tue May 10 17:27:01 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:

I just make a query with some filters and retrieve the corresponding documents. Also, I try to add aggregations about these documents with a `NestedAggregation` and **without any filter** used in the `query` so I used at the very beginning the `GlobalAggregation`.

When I make the request I get a `Null Pointer exception` from elastic log. However, I get results from aggregations but also, an error.

Request:

```
{
    "query": {
        "bool": {
            "filter": [
                {
                    "bool": {
                        "must": [
                            {
                                "nested": {
                                    "path": "attributes",
                                    "query": {
                                        "bool": {
                                            "must": [
                                                {
                                                    "nested": {
                                                        "path": "attributes.value",
                                                        "query": {
                                                            "term": {
                                                                "attributes.value.group_value": "4"
                                                            }
                                                        }
                                                    }
                                                },
                                                {
                                                    "term": {
                                                        "attributes.id": "2"
                                                    }
                                                }
                                            ]
                                        }
                                    }
                                }
                            },
                            {
                                "nested": {
                                    "path": "path",
                                    "query": {
                                        "term": {
                                            "path.id_category": "1"
                                        }
                                    }
                                }
                            }
                        ]
                    }
                }
            ]
        }
    },
    "aggregations": {
        "all_wine_types": {
            "global": {},
            "aggregations": {
                "all_wine_types": {
                    "nested": {
                        "path": "attributes"
                    },
                    "aggregations": {
                        "all_wine_types": {
                            "filter": {
                                "term": {
                                    "attributes.id": "2"
                                }
                            },
                            "aggregations": {
                                "all_wine_type_values": {
                                    "nested": {
                                        "path": "attributes.value"
                                    },
                                    "aggregations": {
                                        "wine_type_value": {
                                            "terms": {
                                                "field": "attributes.value.group_value"
                                            },
                                            "aggregations": {
                                                "data": {
                                                    "reverse_nested": {},
                                                    "aggregations": {
                                                        "data": {
                                                            "top_hits": {
                                                                "size": 1,
                                                                "_source": {
                                                                    "includes": [
                                                                        "attributes"
                                                                    ]
                                                                }
                                                            }
                                                        }
                                                    }
                                                },
                                                "root_data": {
                                                    "reverse_nested": {},
                                                    "aggregations": {
                                                        "data": {
                                                            "top_hits": {
                                                                "size": 1,
                                                                "_source": {
                                                                    "includes": [
                                                                        "path"
                                                                    ]
                                                                }
                                                            }
                                                        }
                                                    }
                                                }
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    },
    "size": 0
}

```

Response:

```
{
    "took": 45,
    "timed_out": false,
    "_shards": {
        "total": 5,
        "successful": 4,
        "failed": 1,
        "failures": [
            {
                "shard": 4,
                "index": "xxxxx",
                "node": "4__8482PTbacWtnUnX4d8g",
                "reason": {
                    "type": "null_pointer_exception",
                    "reason": null
                }
            }
        ]
    },
    "hits": {
        "total": 18544,
        "max_score": 0,
        "hits": []
    },
```

**Steps to reproduce**:

No steps!

**Provide logs (if relevant)**:

```
2017-12-20T15:54:54,361][DEBUG][o.e.a.s.TransportSearchAction] [4__8482] [xxxxx][4], node[4__8482PTbacWtnUnX4d8g], [P], s[STARTED], a[id=LOwrt7QKRK6bdrhhakAkRg]: Failed to execute [SearchRequest{searchType=QUERY_THEN_FETCH, indices=[xxxxx], indicesOptions=IndicesOptions[id=38, ignore_unavailable=false, allow_no_indices=true, expand_wildcards_open=true, expand_wildcards_closed=false, allow_alisases_to_multiple_indices=true, forbid_closed_indices=true], types=[product], routing='null', preference='null', requestCache=null, scroll=null, source={
  "size" : 0,
  "query" : {
    "bool" : {
      "filter" : [
        {
          "bool" : {
            "must" : [
              {
                "nested" : {
                  "query" : {
                    "bool" : {
                      "must" : [
                        {
                          "nested" : {
                            "query" : {
                              "term" : {
                                "attributes.value.group_value" : {
                                  "value" : "4",
                                  "boost" : 1.0
                                }
                              }
                            },
                            "path" : "attributes.value",
                            "ignore_unmapped" : false,
                            "score_mode" : "avg",
                            "boost" : 1.0
                          }
                        },
                        {
                          "term" : {
                            "attributes.id" : {
                              "value" : "2",
                              "boost" : 1.0
                            }
                          }
                        }
                      ],
                      "disable_coord" : false,
                      "adjust_pure_negative" : true,
                      "boost" : 1.0
                    }
                  },
                  "path" : "attributes",
                  "ignore_unmapped" : false,
                  "score_mode" : "avg",
                  "boost" : 1.0
                }
              },
              {
                "nested" : {
                  "query" : {
                    "term" : {
                      "path.id_category" : {
                        "value" : "1",
                        "boost" : 1.0
                      }
                    }
                  },
                  "path" : "path",
                  "ignore_unmapped" : false,
                  "score_mode" : "avg",
                  "boost" : 1.0
                }
              }
            ],
            "disable_coord" : false,
            "adjust_pure_negative" : true,
            "boost" : 1.0
          }
        }
      ],
      "disable_coord" : false,
      "adjust_pure_negative" : true,
      "boost" : 1.0
    }
  },
  "aggregations" : {
    "all_wine_types" : {
      "global" : { },
      "aggregations" : {
        "all_wine_types" : {
          "nested" : {
            "path" : "attributes"
          },
          "aggregations" : {
            "all_wine_types" : {
              "filter" : {
                "term" : {
                  "attributes.id" : {
                    "value" : "2",
                    "boost" : 1.0
                  }
                }
              },
              "aggregations" : {
                "all_wine_type_values" : {
                  "nested" : {
                    "path" : "attributes.value"
                  },
                  "aggregations" : {
                    "wine_type_value" : {
                      "terms" : {
                        "field" : "attributes.value.group_value",
                        "size" : 10,
                        "shard_size" : -1,
                        "min_doc_count" : 1,
                        "shard_min_doc_count" : 0,
                        "show_term_doc_count_error" : false,
                        "order" : [
                          {
                            "_count" : "desc"
                          },
                          {
                            "_term" : "asc"
                          }
                        ]
                      },
                      "aggregations" : {
                        "data" : {
                          "reverse_nested" : { },
                          "aggregations" : {
                            "data" : {
                              "top_hits" : {
                                "from" : 0,
                                "size" : 1,
                                "version" : false,
                                "explain" : false,
                                "_source" : {
                                  "includes" : [
                                    "attributes"
                                  ],
                                  "excludes" : [ ]
                                }
                              }
                            }
                          }
                        },
                        "root_data" : {
                          "reverse_nested" : { },
                          "aggregations" : {
                            "data" : {
                              "top_hits" : {
                                "from" : 0,
                                "size" : 1,
                                "version" : false,
                                "explain" : false,
                                "_source" : {
                                  "includes" : [
                                    "path"
                                  ],
                                  "excludes" : [ ]
                                }
                              }
                            }
                          }
                        }
                      }
                    }
                  }
                }
              }
            }
          }
        }
      }
    }
  },
  "ext" : { }
}}]
org.elasticsearch.transport.RemoteTransportException: [4__8482][192.168.1.12:9300][indices:data/read/search[phase/query]]
Caused by: java.lang.NullPointerException
```

Thank you so much!

--- a/core/src/main/java/org/elasticsearch/search/aggregations/bucket/BestBucketsDeferringCollector.java
+++ b/core/src/main/java/org/elasticsearch/search/aggregations/bucket/BestBucketsDeferringCollector.java
@@ -21,6 +21,8 @@ package org.elasticsearch.search.aggregations.bucket;
 
 import org.apache.lucene.index.LeafReaderContext;
 import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.MatchAllDocsQuery;
+import org.apache.lucene.search.Query;
 import org.apache.lucene.search.Scorer;
 import org.apache.lucene.search.Weight;
 import org.apache.lucene.util.packed.PackedInts;
@@ -59,6 +61,7 @@ public class BestBucketsDeferringCollector extends DeferringBucketCollector {
     final List<Entry> entries = new ArrayList<>();
     BucketCollector collector;
     final SearchContext searchContext;
+    final boolean isGlobal;
     LeafReaderContext context;
     PackedLongValues.Builder docDeltas;
     PackedLongValues.Builder buckets;
@@ -66,9 +69,14 @@ public class BestBucketsDeferringCollector extends DeferringBucketCollector {
     boolean finished = false;
     LongHash selectedBuckets;
 
-    /** Sole constructor. */
-    public BestBucketsDeferringCollector(SearchContext context) {
+    /**
+     * Sole constructor.
+     * @param context The search context
+     * @param isGlobal Whether this collector visits all documents (global context)
+     */
+    public BestBucketsDeferringCollector(SearchContext context, boolean isGlobal) {
         this.searchContext = context;
+        this.isGlobal = isGlobal;
     }
 
     @Override
@@ -144,11 +152,11 @@ public class BestBucketsDeferringCollector extends DeferringBucketCollector {
         }
         this.selectedBuckets = hash;
 
-        boolean needsScores = collector.needsScores();
+        boolean needsScores = needsScores();
         Weight weight = null;
         if (needsScores) {
-            weight = searchContext.searcher()
-                        .createNormalizedWeight(searchContext.query(), true);
+            Query query = isGlobal ? new MatchAllDocsQuery() : searchContext.query();
+            weight = searchContext.searcher().createNormalizedWeight(query, true);
         }
         for (Entry entry : entries) {
             final LeafBucketCollector leafCollector = collector.getLeafCollector(entry.context);

--- a/core/src/main/java/org/elasticsearch/search/aggregations/bucket/DeferableBucketAggregator.java
+++ b/core/src/main/java/org/elasticsearch/search/aggregations/bucket/DeferableBucketAggregator.java
@@ -22,6 +22,7 @@ package org.elasticsearch.search.aggregations.bucket;
 import org.elasticsearch.search.aggregations.Aggregator;
 import org.elasticsearch.search.aggregations.AggregatorFactories;
 import org.elasticsearch.search.aggregations.BucketCollector;
+import org.elasticsearch.search.aggregations.bucket.global.GlobalAggregator;
 import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
 import org.elasticsearch.search.internal.SearchContext;
 
@@ -61,10 +62,20 @@ public abstract class DeferableBucketAggregator extends BucketsAggregator {
         collectableSubAggregators = BucketCollector.wrap(collectors);
     }
 
+    public static boolean descendsFromGlobalAggregator(Aggregator parent) {
+        while (parent != null) {
+            if (parent.getClass() == GlobalAggregator.class) {
+                return true;
+            }
+            parent = parent.parent();
+        }
+        return false;
+    }
+
     public DeferringBucketCollector getDeferringCollector() {
         // Default impl is a collector that selects the best buckets
         // but an alternative defer policy may be based on best docs.
-        return new BestBucketsDeferringCollector(context());
+        return new BestBucketsDeferringCollector(context(), descendsFromGlobalAggregator(parent()));
     }
 
     /**
@@ -74,7 +85,7 @@ public abstract class DeferableBucketAggregator extends BucketsAggregator {
      * recording of all doc/bucketIds from the first pass and then the sub class
      * should call {@link #runDeferredCollections(long...)} for the selected set
      * of buckets that survive the pruning.
-     * 
+     *
      * @param aggregator
      *            the child aggregator
      * @return true if the aggregator should be deferred until a first pass at

--- a/core/src/test/java/org/elasticsearch/search/aggregations/bucket/BestBucketsDeferringCollectorTests.java
+++ b/core/src/test/java/org/elasticsearch/search/aggregations/bucket/BestBucketsDeferringCollectorTests.java
@@ -27,6 +27,8 @@ import org.apache.lucene.index.LeafReaderContext;
 import org.apache.lucene.index.RandomIndexWriter;
 import org.apache.lucene.index.Term;
 import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.MatchAllDocsQuery;
+import org.apache.lucene.search.Query;
 import org.apache.lucene.search.ScoreDoc;
 import org.apache.lucene.search.TermQuery;
 import org.apache.lucene.search.TopDocs;
@@ -40,6 +42,8 @@ import java.io.IOException;
 import java.util.Collections;
 import java.util.HashSet;
 import java.util.Set;
+
+import static org.mockito.Mockito.when;
 
 public class BestBucketsDeferringCollectorTests extends AggregatorTestCase {
 
@@ -59,14 +63,35 @@ public class BestBucketsDeferringCollectorTests extends AggregatorTestCase {
         IndexSearcher indexSearcher = new IndexSearcher(indexReader);
 
         TermQuery termQuery = new TermQuery(new Term("field", String.valueOf(randomInt(maxNumValues))));
+        Query rewrittenQuery = indexSearcher.rewrite(termQuery);
         TopDocs topDocs = indexSearcher.search(termQuery, numDocs);
 
         SearchContext searchContext = createSearchContext(indexSearcher, createIndexSettings());
-        BestBucketsDeferringCollector collector = new BestBucketsDeferringCollector(searchContext);
+        when(searchContext.query()).thenReturn(rewrittenQuery);
+        BestBucketsDeferringCollector collector = new BestBucketsDeferringCollector(searchContext, false) {
+            @Override
+            public boolean needsScores() {
+                return true;
+            }
+        };
         Set<Integer> deferredCollectedDocIds = new HashSet<>();
         collector.setDeferredCollector(Collections.singleton(bla(deferredCollectedDocIds)));
         collector.preCollection();
         indexSearcher.search(termQuery, collector);
+        collector.postCollection();
+        collector.replay(0);
+
+        assertEquals(topDocs.scoreDocs.length, deferredCollectedDocIds.size());
+        for (ScoreDoc scoreDoc : topDocs.scoreDocs) {
+            assertTrue("expected docid [" + scoreDoc.doc + "] is missing", deferredCollectedDocIds.contains(scoreDoc.doc));
+        }
+
+        topDocs = indexSearcher.search(new MatchAllDocsQuery(), numDocs);
+        collector = new BestBucketsDeferringCollector(searchContext, true);
+        deferredCollectedDocIds = new HashSet<>();
+        collector.setDeferredCollector(Collections.singleton(bla(deferredCollectedDocIds)));
+        collector.preCollection();
+        indexSearcher.search(new MatchAllDocsQuery(), collector);
         collector.postCollection();
         collector.replay(0);
 

--- a/core/src/test/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorTests.java
+++ b/core/src/test/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorTests.java
@@ -46,16 +46,20 @@ import org.elasticsearch.index.mapper.MappedFieldType;
 import org.elasticsearch.index.mapper.NumberFieldMapper;
 import org.elasticsearch.index.query.QueryBuilders;
 import org.elasticsearch.indices.breaker.NoneCircuitBreakerService;
+import org.elasticsearch.search.SearchHit;
 import org.elasticsearch.search.aggregations.AggregationBuilder;
 import org.elasticsearch.search.aggregations.AggregationBuilders;
 import org.elasticsearch.search.aggregations.Aggregator;
 import org.elasticsearch.search.aggregations.AggregatorTestCase;
 import org.elasticsearch.search.aggregations.BucketOrder;
 import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.InternalMultiBucketAggregation;
+import org.elasticsearch.search.aggregations.bucket.MultiBucketsAggregation;
 import org.elasticsearch.search.aggregations.bucket.filter.Filter;
 import org.elasticsearch.search.aggregations.bucket.filter.FilterAggregationBuilder;
 import org.elasticsearch.search.aggregations.bucket.global.GlobalAggregationBuilder;
 import org.elasticsearch.search.aggregations.bucket.global.InternalGlobal;
+import org.elasticsearch.search.aggregations.metrics.tophits.InternalTopHits;
 import org.elasticsearch.search.aggregations.metrics.tophits.TopHitsAggregationBuilder;
 import org.elasticsearch.search.aggregations.support.ValueType;
 
@@ -71,6 +75,7 @@ import java.util.function.BiFunction;
 import java.util.function.Function;
 
 import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
 import static org.hamcrest.Matchers.instanceOf;
 
 public class TermsAggregatorTests extends AggregatorTestCase {
@@ -961,7 +966,7 @@ public class TermsAggregatorTests extends AggregatorTestCase {
                                 .field("keyword")
                                 .order(BucketOrder.key(true))
                                 .subAggregation(
-                                    new TermsAggregationBuilder("_name2", ValueType.STRING)
+                                    new TermsAggregationBuilder("sub_terms", ValueType.STRING)
                                         .executionHint(executionHint)
                                         .collectMode(collectionMode)
                                         .field("keyword").order(BucketOrder.key(true))
@@ -977,8 +982,18 @@ public class TermsAggregatorTests extends AggregatorTestCase {
                     fieldType.setHasDocValues(true);
 
                     InternalGlobal result = searchAndReduce(indexSearcher, new MatchAllDocsQuery(), globalBuilder, fieldType);
-                    Terms terms = result.getAggregations().get("terms");
+                    InternalMultiBucketAggregation<?, ?> terms = result.getAggregations().get("terms");
                     assertThat(terms.getBuckets().size(), equalTo(3));
+                    for (MultiBucketsAggregation.Bucket bucket : terms.getBuckets()) {
+                        InternalMultiBucketAggregation<?, ?> subTerms = bucket.getAggregations().get("sub_terms");
+                        assertThat(subTerms.getBuckets().size(), equalTo(1));
+                        MultiBucketsAggregation.Bucket subBucket  = subTerms.getBuckets().get(0);
+                        InternalTopHits topHits = subBucket.getAggregations().get("top_hits");
+                        assertThat(topHits.getHits().getHits().length, equalTo(1));
+                        for (SearchHit hit : topHits.getHits()) {
+                            assertThat(hit.getScore(), greaterThan(0f));
+                        }
+                    }
                 }
             }
         }

--- a/core/src/test/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorTests.java
+++ b/core/src/test/java/org/elasticsearch/search/aggregations/bucket/terms/TermsAggregatorTests.java
@@ -54,6 +54,9 @@ import org.elasticsearch.search.aggregations.BucketOrder;
 import org.elasticsearch.search.aggregations.InternalAggregation;
 import org.elasticsearch.search.aggregations.bucket.filter.Filter;
 import org.elasticsearch.search.aggregations.bucket.filter.FilterAggregationBuilder;
+import org.elasticsearch.search.aggregations.bucket.global.GlobalAggregationBuilder;
+import org.elasticsearch.search.aggregations.bucket.global.InternalGlobal;
+import org.elasticsearch.search.aggregations.metrics.tophits.TopHitsAggregationBuilder;
 import org.elasticsearch.search.aggregations.support.ValueType;
 
 import java.io.IOException;
@@ -67,6 +70,7 @@ import java.util.Map;
 import java.util.function.BiFunction;
 import java.util.function.Function;
 
+import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.instanceOf;
 
 public class TermsAggregatorTests extends AggregatorTestCase {
@@ -929,6 +933,53 @@ public class TermsAggregatorTests extends AggregatorTestCase {
                 assertEquals(expected * 2, buckets.get(2).getDocCount());
                 assertEquals("1000.0", buckets.get(3).getKeyAsString());
                 assertEquals(expected, buckets.get(3).getDocCount());
+            }
+        }
+    }
+
+    public void testGlobalAggregationWithScore() throws IOException {
+        try (Directory directory = newDirectory()) {
+            try (RandomIndexWriter indexWriter = new RandomIndexWriter(random(), directory)) {
+                Document document = new Document();
+                document.add(new SortedDocValuesField("keyword", new BytesRef("a")));
+                indexWriter.addDocument(document);
+                document = new Document();
+                document.add(new SortedDocValuesField("keyword", new BytesRef("c")));
+                indexWriter.addDocument(document);
+                document = new Document();
+                document.add(new SortedDocValuesField("keyword", new BytesRef("e")));
+                indexWriter.addDocument(document);
+                try (IndexReader indexReader = maybeWrapReaderEs(indexWriter.getReader())) {
+                    IndexSearcher indexSearcher = newIndexSearcher(indexReader);
+                    String executionHint = randomFrom(TermsAggregatorFactory.ExecutionMode.values()).toString();
+                    Aggregator.SubAggCollectionMode collectionMode = randomFrom(Aggregator.SubAggCollectionMode.values());
+                    GlobalAggregationBuilder globalBuilder = new GlobalAggregationBuilder("global")
+                        .subAggregation(
+                            new TermsAggregationBuilder("terms", ValueType.STRING)
+                                .executionHint(executionHint)
+                                .collectMode(collectionMode)
+                                .field("keyword")
+                                .order(BucketOrder.key(true))
+                                .subAggregation(
+                                    new TermsAggregationBuilder("_name2", ValueType.STRING)
+                                        .executionHint(executionHint)
+                                        .collectMode(collectionMode)
+                                        .field("keyword").order(BucketOrder.key(true))
+                                        .subAggregation(
+                                            new TopHitsAggregationBuilder("top_hits")
+                                                .storedField("_none_")
+                                        )
+                                )
+                        );
+
+                    MappedFieldType fieldType = new KeywordFieldMapper.KeywordFieldType();
+                    fieldType.setName("keyword");
+                    fieldType.setHasDocValues(true);
+
+                    InternalGlobal result = searchAndReduce(indexSearcher, new MatchAllDocsQuery(), globalBuilder, fieldType);
+                    Terms terms = result.getAggregations().get("terms");
+                    assertThat(terms.getBuckets().size(), equalTo(3));
+                }
             }
         }
     }

